Base backups Postgres 10

Hi,

I'm attempting to use base backups on Postgres 10, and am receiving an error when running the tests:

```
error: unexpected response: &{%!t(string=ERROR) %!t(string=42601) %!t(string=syntax error) %!t(string=) %!t(string=) %!t(int32=0) %!t(int32=0) %!t(string=) %!t(string=) %!t(string=) %!t(string=) %!t(string=) %!t(string=) %!t(string=) %!t(string=repl_scanner.l) %!t(int32=217) %!t(string=replication_yyerror) map[%!t(uint8=86):%!t(string=ERROR)]}
--- FAIL: TestBaseBackup (0.01s)
    pglogrepl_test.go:351:
        	Error Trace:	pglogrepl_test.go:351
        	Error:      	"0" is not greater than or equal to "1"
        	Test:       	TestBaseBackup
FAIL
exit status 1
FAIL	github.com/jackc/pglogrepl	0.812s
```

which essentially translates to:

```
&{ERROR 42601 syntax error   %!s(int32=0) %!s(int32=0)        repl_scanner.l %!s(int32=217) replication_yyerror map[%!s(uint8=86):ERROR]}
```

the error is being thrown by `getBaseBackupInfo` so I suspect something in the options is not compatible with pg10.

I'll update the ticket as I find more.

--- a/README.md
+++ b/README.md
@@ -32,6 +32,12 @@ Create a user:
 create user pglogrepl with replication password 'secret';
 ```
 
+If you're using PostgreSQL 15 or newer grant access to the public schema, just for these tests:
+
+```
+grant all on schema public to pglogrepl;
+```
+
 Add a replication line to your pg_hba.conf:
 
 ```

--- a/pglogrepl.go
+++ b/pglogrepl.go
@@ -444,6 +444,8 @@ func getBaseBackupInfo(ctx context.Context, conn *pgconn.PgConn) (start LSN, tim
 		case *pgproto3.NoticeResponse:
 		case *pgproto3.CommandComplete:
 			return start, timelineID, nil
+		case *pgproto3.ErrorResponse:
+			return start, timelineID, errors.Errorf("error response sev=%q code=%q message=%q detail=%q position=%d", msg.Severity, msg.Code, msg.Message, msg.Detail, msg.Position)
 		default:
 			return start, timelineID, errors.Errorf("unexpected response: %t", msg)
 		}

--- a/pglogrepl.go
+++ b/pglogrepl.go
@@ -344,8 +344,8 @@ type BaseBackupOptions struct {
 	NoVerifyChecksums bool
 }
 
-func (bbo BaseBackupOptions) sql() string {
-	parts := []string{"BASE_BACKUP"}
+func (bbo BaseBackupOptions) sql(serverVersion int) string {
+	parts := []string{}
 	if bbo.Label != "" {
 		parts = append(parts, "LABEL '"+strings.ReplaceAll(bbo.Label, "'", "''")+"'")
 	}
@@ -353,13 +353,21 @@ func (bbo BaseBackupOptions) sql() string {
 		parts = append(parts, "PROGRESS")
 	}
 	if bbo.Fast {
-		parts = append(parts, "FAST")
+		if serverVersion >= 15 {
+			parts = append(parts, "CHECKPOINT 'fast'")
+		} else {
+			parts = append(parts, "FAST")
+		}
 	}
 	if bbo.WAL {
 		parts = append(parts, "WAL")
 	}
 	if bbo.NoWait {
-		parts = append(parts, "NOWAIT")
+		if serverVersion >= 15 {
+			parts = append(parts, "WAIT false")
+		} else {
+			parts = append(parts, "NOWAIT")
+		}
 	}
 	if bbo.MaxRate >= 32 {
 		parts = append(parts, fmt.Sprintf("MAX_RATE %d", bbo.MaxRate))
@@ -368,9 +376,16 @@ func (bbo BaseBackupOptions) sql() string {
 		parts = append(parts, "TABLESPACE_MAP")
 	}
 	if bbo.NoVerifyChecksums {
-		parts = append(parts, "NOVERIFY_CHECKSUMS")
+		if serverVersion >= 15 {
+			parts = append(parts, "VERIFY_CHECKSUMS false")
+		} else if serverVersion >= 11 {
+			parts = append(parts, "NOVERIFY_CHECKSUMS")
+		}
 	}
-	return strings.Join(parts, " ")
+	if serverVersion >= 15 {
+		return "BASE_BACKUP(" + strings.Join(parts, ", ") + ")"
+	}
+	return "BASE_BACKUP " + strings.Join(parts, " ")
 }
 
 // BaseBackupTablespace represents a tablespace in the backup
@@ -387,9 +402,22 @@ type BaseBackupResult struct {
 	Tablespaces []BaseBackupTablespace
 }
 
+func serverMajorVersion(conn *pgconn.PgConn) (int, error) {
+	verString := conn.ParameterStatus("server_version")
+	dot := strings.IndexByte(verString, '.')
+	if dot == -1 {
+		return 0, errors.Errorf("bad server version string: '%s'", verString)
+	}
+	return strconv.Atoi(verString[:dot])
+}
+
 // StartBaseBackup begins the process for copying a basebackup by executing the BASE_BACKUP command.
 func StartBaseBackup(ctx context.Context, conn *pgconn.PgConn, options BaseBackupOptions) (result BaseBackupResult, err error) {
-	sql := options.sql()
+	serverVersion, err := serverMajorVersion(conn)
+	if err != nil {
+		return result, err
+	}
+	sql := options.sql(serverVersion)
 
 	buf := (&pgproto3.Query{String: sql}).Encode(nil)
 	err = conn.SendBytes(ctx, buf)

--- a/pglogrepl_test.go
+++ b/pglogrepl_test.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"os"
 	"strconv"
+	"strings"
 	"testing"
 	"time"
 
@@ -336,8 +337,10 @@ func TestBaseBackup(t *testing.T) {
 	require.NoError(t, err)
 	defer closeConn(t, conn)
 
+	// NOVERIFY_CHECKSUMS is only supported in PG11 and later
+	noVerifyChecksums := !strings.HasPrefix(conn.ParameterStatus("server_version"), "10.")
 	options := pglogrepl.BaseBackupOptions{
-		NoVerifyChecksums: true,
+		NoVerifyChecksums: noVerifyChecksums,
 		Progress:          true,
 		Label:             "pglogrepltest",
 		Fast:              true,
@@ -347,8 +350,8 @@ func TestBaseBackup(t *testing.T) {
 		TablespaceMap:     true,
 	}
 	startRes, err := pglogrepl.StartBaseBackup(context.Background(), conn, options)
-	require.GreaterOrEqual(t, startRes.TimelineID, int32(1))
 	require.NoError(t, err)
+	require.GreaterOrEqual(t, startRes.TimelineID, int32(1))
 
 	//Write the tablespaces
 	for i := 0; i < len(startRes.Tablespaces)+1; i++ {

--- a/pglogrepl_test.go
+++ b/pglogrepl_test.go
@@ -5,7 +5,6 @@ import (
 	"fmt"
 	"os"
 	"strconv"
-	"strings"
 	"testing"
 	"time"
 
@@ -225,11 +224,18 @@ drop table t;
 	}
 
 	rxXLogData := func() pglogrepl.XLogData {
-		msg, err := conn.ReceiveMessage(ctx)
-		require.NoError(t, err)
-		cdMsg, ok := msg.(*pgproto3.CopyData)
-		require.True(t, ok)
-
+		var cdMsg *pgproto3.CopyData
+		// Discard keepalive messages
+		for {
+			msg, err := conn.ReceiveMessage(ctx)
+			require.NoError(t, err)
+			var ok bool
+			cdMsg, ok = msg.(*pgproto3.CopyData)
+			require.True(t, ok)
+			if cdMsg.Data[0] != pglogrepl.PrimaryKeepaliveMessageByteID {
+				break
+			}
+		}
 		require.Equal(t, byte(pglogrepl.XLogDataByteID), cdMsg.Data[0])
 		xld, err := pglogrepl.ParseXLogData(cdMsg.Data[1:])
 		require.NoError(t, err)
@@ -337,10 +343,8 @@ func TestBaseBackup(t *testing.T) {
 	require.NoError(t, err)
 	defer closeConn(t, conn)
 
-	// NOVERIFY_CHECKSUMS is only supported in PG11 and later
-	noVerifyChecksums := !strings.HasPrefix(conn.ParameterStatus("server_version"), "10.")
 	options := pglogrepl.BaseBackupOptions{
-		NoVerifyChecksums: noVerifyChecksums,
+		NoVerifyChecksums: true,
 		Progress:          true,
 		Label:             "pglogrepltest",
 		Fast:              true,

